### PR TITLE
Add typing to ``filepath``

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

This fails the test suite, but I think this might be because the test suite is testing something that doesn't happen in normal use of pylint. Please correct me if I am wrong but after investigating the call-chain I believe that almost all `filepath`'s handled by `pylint` are `str` (the only exception being `get_fatal_error_message()`). This presumes that no user directly calls `linter.check()` as we do in the test suite. 
Of note is `lint.py_run()`, which I believe also does no run `check()` before calling `load_command_line_configuration`.

The following reasoning leads me to believe that a normal use of `pylint` always creates a `str` file path.
- `load_command_line_configuration()` in `pylint/config/option_manager_mixin.py` always returns a `List[str]`. Therefore the `args` variable in the `__init__` of the `Run` class in `pylint/lint/run.py` is always a `List[str]`.
- Subsequently in `pylint/lint/pylinter.py` the `files_or_modules` argument to `check()` is always a `List[str]`. Which means that both `check_parallel()` and `_check_files()` only handle `str` representations of paths.
- Therefore `set_current_module()` and `on_set_current_module()` in various other classes only get given `str`.

https://github.com/DanielNoord/pylint/pull/10
This PR shows that with changes to the tests this PR can succeed. These changes are:
- Making sure all initial file paths are a `List[str]`
- Making sure that `test_worker_check_single_file_uninitialised()` catches the new explicitly defined exception
- Making sure that `get_fatal_error_message()` always gets a `Path` as second argument
- Making sure that `check_parallel()` always gets an `Iterable`

These changes seem to follow the standard format of these arguments/variables in normal runs of `pylint` that are being skipped in the tests.

### CODE REVIEW

File: pylint/lint/expand_modules.py
Comment (DanielNoord) [inline pylint/lint/expand_modules.py:13]: After a (potential) merge of #4973 these should be added to `pylint/typing`.
This PR is therefore blocked by #4973

File: pylint/lint/parallel.py
Comment (DanielNoord) [inline pylint/lint/parallel.py:72]: `mpy` complains on the following two lines about the possibility that `_work_linter` might be None.
This is probably the wrong `Exception` type.
We currently already test for this, but do so with catching an `AttributeError` on L73. See:
https://github.com/PyCQA/pylint/blob/fab2de272b3fc36dc4480d89840b8bfe0fbfbd0c/tests/test_check_parallel.py#L177-L182
I think making this error more explicit as `mypy` wants makes sense.

File: pylint/lint/expand_modules.py
Comment (cdce8p) [inline pylint/lint/expand_modules.py:13]: This should use `PascalCase`. Same for `ERROR_DESCRIPTION_DICT` below.

File: pylint/lint/parallel.py
Comment (cdce8p) [inline pylint/lint/parallel.py:67]: Does it make sense to define the type for `file_item` as NamedTuple? It seems to be used in a few places.

**Review by cdce8p — COMMENTED** [review]
Not a review, just some thoughts.

File: pylint/lint/parallel.py
Comment (cdce8p) [inline pylint/lint/parallel.py:159]: Mentioning line numbers in comments isn't worth it. The code will eventually get updated without the comment.

File: pylint/lint/parallel.py
Comment (DanielNoord) [reply pylint/lint/parallel.py:67]: Good call!
Probably after https://github.com/PyCQA/pylint/pull/4973 though to add it to the `typing.py` document.

File: pylint/lint/parallel.py
Comment (cdce8p) [inline pylint/lint/parallel.py:159]:
```suggestion
                linter.reporter.handle_message(msg)  # type: ignore  # linter.set_reporter() call above makes linter have a reporter attr
```

**Review by cdce8p — COMMENTED** [review]
I wouldn't necessarily wait for #4973 to add `typing.py`. You could also add it here. It shouldn't be too hard to resolve the merge conflicts in the end.

For `Tuple[str, str, str]` it might even make sense to create a separate one. You should you by now that I like small PRs 😄

**Comment by DanielNoord** [comment]
@cdce8p I added the NamedTuple. However, doing this separately was (quite) awful because of the various new `mypy` errors that I already fixed in this PR. (For example raising the Exception on `if not _worker_linter`).
So I added it here!

File: pylint/lint/expand_modules.py
Comment (cdce8p) [inline pylint/lint/expand_modules.py:24]: You can move these to `pylint/typing.py` now.

File: pylint/lint/parallel.py
Comment (cdce8p) [inline pylint/lint/parallel.py:118]: Does it need to be an `Iterator` or is `Iterable` enough?
https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:940]: `Iterable`?

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:972]: These arguments seems to represent a `FileItem`.

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:987]:
```suggestion
        file_descrs: Iterable[FileItem],
```

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:988]:
```suggestion
    ) -> None:
```

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:1016]: These again seem to represent a `FileItem`.

**Review by cdce8p — COMMENTED** [review]
I haven't commented on that before. A best practice when typing in python is to use the most generic type possible for arguments and the most concrete one for return types. I.e.
```py
def func(var: Iterable[str]) -> List[str]:
    ...
```

If it doesn't matter if a `list`, `tuple` or something else is pass in as long as it's possible to iterate over it, it shouldn't matter to the type checker either.

File: pylint/lint/pylinter.py
Comment (DanielNoord) [reply pylint/lint/pylinter.py:972]: Needed some refactoring, but added!

File: pylint/lint/pylinter.py
Comment (DanielNoord) [reply pylint/lint/pylinter.py]: Done in own commit!

File: pylint/lint/parallel.py
Comment (cdce8p) [inline pylint/lint/parallel.py:86]:
```suggestion
        file_item.filepath,
```
One advantage of using a NamedTuple.

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:940]: `Sequence` is also an excellent choice!

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:943]:
```suggestion
        files_or_modules is a sequence of strings presenting modules to check.
```

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:1023]: Why do a tuple unpacking. With `NamedTuple` you can access the attributes by name.
It might even make sense to just pass `file` to `set_current_module` / `get_ast` and do the attribute access there. A bit more refactoring though.

**Review by cdce8p — COMMENTED** [review]
I really like where this is going!

**Comment by Pierre-Sassoulas** [comment]
>  I believe that almost all filepath's handled by pylint are str

We should handle ``Union[Path, str]`` imo.

File: pylint/lint/pylinter.py
Comment (DanielNoord) [reply pylint/lint/pylinter.py]: I left the refactor for another time, since this is becoming quite a large PR.

One thing I noticed is some ambiguity between `modname` and `modpath`. They seem to be used for the same attribute. We currently use `modpath`, is that okay? Or do you want to start using `modname`?

Comment (Pierre-Sassoulas) [reply pylint/lint/pylinter.py]: +1 for keeping the refactors for later :)

**Comment by DanielNoord** [comment]
> We should handle `Union[Path, str]` imo.

Not sure if I agree here. This would probably lead us to having to add `Path` to many other typings down the line. The current flow of `pylint` clearly does not allow `Path` to be passed. With a little refactor we can even get the last instance of `Path` to also become a `str`. Having to add `Union[Path, str]` instead of `str` everywhere might clutter the codebase.

@cdce8p & @Pierre-Sassoulas Do you guys thinks your review is far enough for me to add the commit with test fixes as discussed in the PR description and shown in https://github.com/DanielNoord/pylint/pull/10?

**Comment by Pierre-Sassoulas** [comment]
> might clutter the codebase.

Yes, also the migration to Path is also a huge endeavor with not a lot of clear benefit right now. Most ``os.path`` handles ``Union[Path, str]`` so technically that's what we can probably handle on most of the function, but finding out where we can't is probably not worth the time and crashes it will create.

**Comment by DanielNoord** [comment]
This now includes the relevant changes to the tests to make this work. As discussed in PR description I believe the tests check things that cannot happen in real world scenarios. The description includes some additional reasoning for this and a description of the changes to the tests.

File: pylint/lint/pylinter.py
Comment (Pierre-Sassoulas) [inline pylint/lint/pylinter.py:950]: I like this, clearer input, it will make the code better. But I think this might be a breaking change.

File: pylint/lint/pylinter.py
Comment (Pierre-Sassoulas) [inline pylint/lint/pylinter.py:1060]: 👍

File: pylint/lint/pylinter.py
Comment (cdce8p) [reply pylint/lint/pylinter.py:950]: Technically all changes of function arguments are breaking ones as well (`check_single_file`, `check`, `_check_files`). It depends if we would consider them part of the public API. Something I'm not sure about.

Comment (Pierre-Sassoulas) [reply pylint/lint/pylinter.py:950]: Well I think the check function can be used by addon relying on pylint's checker. I'm for changing the typing only right now and removing the mechanism in 3.0 as a breaking change (we can also add a deprecation warning).

File: pylint/lint/pylinter.py
Comment (Pierre-Sassoulas) [inline pylint/lint/pylinter.py:948]:
```suggestion
        if not isinstance(files_or_modules, (list, tuple)):
            warnins.warn("In pylint 3.0, the checkers check function will only accept sequence of string")
            files_or_modules = (files_or_modules,)
```

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:942]: `Union[Sequence[str], str]` maybe with a `TODO` comment after the deprecation warning to update the type annotation.

File: pylint/lint/pylinter.py
Comment (DanielNoord) [reply pylint/lint/pylinter.py:1019]: As in, keep the `FileItem` or revert to previous state?

Comment (cdce8p) [reply pylint/lint/pylinter.py:1018]: 😅 Keep the change with `FileItem`

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:945]: This should be reverted

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:949]:
```suggestion
            # pylint: disable-next=fixme
            # TODO: Update typing and docstring for 'files_or_modules' when removing the deprecation
            warnings.warn(
```

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:951]:
```suggestion
                DeprecationWarning,
```

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:981]:
```suggestion
            "In pylint 3.0, the checkers check_single_file function will be removed. "
            "Use check_single_file_item instead.",
```
What do you think?

File: pylint/lint/pylinter.py
Comment (cdce8p) [inline pylint/lint/pylinter.py:982]:
```suggestion
            DeprecationWarning,
```

File: pylint/typing.py
Comment (cdce8p) [inline pylint/typing.py:23]:
```suggestion
    - modname: module name
    """
```
PEP8 -> closing quotes should be on a new line if docstring > 1 line.

File: pylint/lint/pylinter.py
Comment (DanielNoord) [reply pylint/lint/pylinter.py:981]: Don't really have a preference. Guess this works as well and requires less refactoring for `3.0`.

File: pylint/lint/pylinter.py
Comment (DanielNoord) [reply pylint/lint/pylinter.py]: Weird how Github doesn't allow a suggestion here..

Comment (cdce8p) [reply pylint/lint/pylinter.py]: I've highlighted both an old an a new line. That's probably why.

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**
